### PR TITLE
fix fetch-incidents returns duplicates bug

### PR DESCRIPTION
**Pull request by YaelShamai**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/XSUP-39390)

## Description
Fixed a bug in the fetch-incidents command which used to return duplicate alerts.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 1]
```diff
@@ -1315,11 +1311,24 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
         ),
         limit=fetch_limit,
     )
+
+    # Some alerts have exactly the same backend_timestamp,
+    # therefor we dedup alerts that have the backend_timestamp of the last alert in the last run.
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1316]:
```suggestion
   # Alerts may be created with the same backend_timestamp. Therefore, we deduplicate alerts that have the same # backend_timestamp as the last alert we saved in the previous run.
```

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 2]
```diff
@@ -1315,11 +1311,24 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
...
+    def get_backend_timestamp(alert):
+        return alert['backend_timestamp']
+    alert_ids_grouped_by_backend_timestamp = map_reduce(response['results'], get_backend_timestamp)
+    # backend_timestamp of last alert retrieved.
+    last_backend_timestamp = response['results'][-1]['backend_timestamp']
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1321]: I don't like the "backend" in the variable name, maybe change to "last_fetched_alert_timestamp".

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 3]
```diff
@@ -1315,11 +1311,24 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
         ),
         limit=fetch_limit,
     )
+
+    # Some alerts have exactly the same backend_timestamp,
+    # therefor we dedup alerts that have the backend_timestamp of the last alert in the last run.
+    def get_backend_timestamp(alert):
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1317]: I don't like one-liner functions.

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 4]
```diff
@@ -1315,11 +1311,24 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
...
+        return alert['backend_timestamp']
+    alert_ids_grouped_by_backend_timestamp = map_reduce(response['results'], get_backend_timestamp)
+    # backend_timestamp of last alert retrieved.
+    last_backend_timestamp = response['results'][-1]['backend_timestamp']
+    # alert ids that have that backend_timestamp (need to be deduped in the next run).
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1322]:
```suggestion
    # All IDs of alerts that share the same timestamp as the last one.
```

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.yml [hunk 5]
```diff
@@ -65,6 +65,7 @@ configuration:
   type: 0
   section: Collect
   required: false
+  additionalinfo: Note, due to duplicates alert that the API retreives, there might be less alerts than the set limit.
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.yml:68]:
```suggestion
  additionalinfo: Note: If the API returns duplicates, we will drop them, causing the total number of alerts fetched to potentially be fewer than the set limit.
```

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 6]
```diff
@@ -1292,17 +1292,13 @@ def get_file_path_command(client: Client, args: dict) -> CommandResults:
 
 
 def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run: dict) -> tuple[list, dict]:
-    # The new API version (v7) always returns the previous last alert along with the new alerts.
-    if not (int_fetch_limit := arg_to_number(fetch_limit)):
-        raise ValueError("limit cannot be empty.")
-    if last_run:
-        int_fetch_limit += 1
+
+    from more_itertools import map_reduce
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1296]: this should be on the top of the file

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 7]
```diff
@@ -1292,17 +1292,13 @@ def get_file_path_command(client: Client, args: dict) -> CommandResults:
...
+
     last_fetched_alert_create_time = last_run.get('last_fetched_alert_create_time')
-    last_fetched_alert_id = last_run.get('last_fetched_alert_id', '')
+    last_fetched_alerts_ids = last_run.get('last_fetched_alerts_ids', '')
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1299]:
```suggestion
    last_fetched_alerts_ids = last_run.get('last_fetched_alerts_ids', [])
```

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 8]
```diff
@@ -1315,11 +1311,24 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
         ),
         limit=fetch_limit,
     )
+
+    # Some alerts have exactly the same backend_timestamp,
+    # therefor we dedup alerts that have the backend_timestamp of the last alert in the last run.
+    def get_backend_timestamp(alert):
+        return alert['backend_timestamp']
+    alert_ids_grouped_by_backend_timestamp = map_reduce(response['results'], get_backend_timestamp)
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1319]: What happens if the response is empty?

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py (hunk 2 above)
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1321]: here too:
What happens if the response is empty?

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 9]
```diff
@@ -1315,11 +1311,24 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
...
+    # backend_timestamp of last alert retrieved.
+    last_backend_timestamp = response['results'][-1]['backend_timestamp']
+    # alert ids that have that backend_timestamp (need to be deduped in the next run).
+    alerts_ids_to_save = [alert['id'] for alert in alert_ids_grouped_by_backend_timestamp[last_backend_timestamp]]
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1323]: here too

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py [hunk 10]
```diff
@@ -1332,11 +1341,9 @@ def fetch_incidents(client: Client, fetch_time: str, fetch_limit: str, last_run:
         incidents.append(incident)
         parsed_date = dateparser.parse(alert_create_date)
         assert parsed_date is not None, f'failed parsing {alert_create_date}'
-        latest_alert_create_date = datetime.strftime(parsed_date + timedelta(seconds=1),
-                                                     '%Y-%m-%dT%H:%M:%S.000Z')
-        latest_alert_id = alert_id
+
     demisto.debug(f'{LOG_INIT} sending {len(incidents)} incidents')
-    res = {'last_fetched_alert_create_time': latest_alert_create_date, 'last_fetched_alert_id': latest_alert_id}
+    res = {'last_fetched_alert_create_time': alert_create_date, 'last_fetched_alerts_ids': alerts_ids_to_save}
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.py:1346]: Make sure if the response was empty, i don't want the last run to change

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR_test.py [hunk 11]
```diff
@@ -439,3 +439,71 @@ def test_alert_process_search_command_func_called(mocker, args, func_to_be_calle
...
+
+
+@pytest.mark.parametrize('response, last_run, expected_incidents', test_fetch_incidents_second_run_data)
+def test_fetch_incidents_second_run(mocker, response, last_run, expected_incidents):
+    """
+    Given:
+        - All arguments needed.
+
+    When:
+        - Running 'fetch-incidents' command in the second time (there is a last_run value).
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR_test.py:497]:
```suggestion
        - When the fetch is running for the second cycle (there is an existing last_run to use).
```

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR_test.py [hunk 12]
```diff
@@ -439,3 +439,71 @@ def test_alert_process_search_command_func_called(mocker, args, func_to_be_calle
...
+
+    Then:
+        - The fetch_incidents fun returns the right incidents.
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR_test.py:500]: Please add an explanation: "should drop the duplicates and return only these".
please note the typo "fun"

File: Packs/CarbonBlackEnterpriseEDR/ReleaseNotes/1_1_35.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### Carbon Black Enterprise EDR
+
+- Fixed a bug in the **fetch-incidents** which used to return duplicate alerts.
```
Comment (RosenbergYehuda) [inline Packs/CarbonBlackEnterpriseEDR/ReleaseNotes/1_1_35.md:6]:
```suggestion
- Fixed an issue in the **fetch-incidents** which used to return duplicate alerts.
```

**Review by RosenbergYehuda — CHANGES_REQUESTED** [review]

**Comment by RosenbergYehuda** [comment]
Great job @YaelShamai !!

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.yml [hunk 14]
```diff
@@ -65,14 +65,15 @@ configuration:
   type: 0
   section: Collect
   required: false
+  additionalinfo: Note, If the API returns duplicates, we will drop them, causing the total number of alerts fetched to potentially be fewer than the set limit.
```
Comment (ShirleyDenkberg) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/CarbonBlackEnterpriseEDR.yml:68]:
```suggestion
  additionalinfo: If the API returns duplicates, the duplicates will be dropped, causing the total number of alerts fetched to potentially be fewer than the set limit.
```

File: Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/README.md [hunk 15]
```diff
@@ -2971,3 +2971,9 @@ Creates a process search job and returns results if 'polling' argument is True.
 | CarbonBlackEEDR.SearchProcess.results.process_username | String | User names related to the process. | 
 | CarbonBlackEEDR.SearchProcess.results.regmod_count | Number | The cumulative count of registry modifications since process tracking started. | 
 | CarbonBlackEEDR.SearchProcess.results.scriptload_count | Number | The cumulative count of loaded scripts since process tracking started. | 
+
+
+## Troubleshooting
+
+Note, If the API returns duplicates, we will drop them, causing the total number of alerts fetched to potentially be fewer than the set limit.
```
Comment (ShirleyDenkberg) [inline Packs/CarbonBlackEnterpriseEDR/Integrations/CarbonBlackEnterpriseEDR/README.md:2978]:
```suggestion
Note: If the API returns duplicates, the duplicates will be dropped, causing the total number of alerts fetched to potentially be fewer than the set limit.
```
This is really not a troubleshooting issue. Not sure it belongs under a Troubleshooting heading.  It's just a Note.  Perhaps it belongs with **fetch-incidents**.

File: Packs/CarbonBlackEnterpriseEDR/ReleaseNotes/1_1_35.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### Carbon Black Enterprise EDR
+
+- Fixed an issue in the **fetch-incidents** which used to return duplicate alerts.
```
Comment (ShirleyDenkberg) [inline Packs/CarbonBlackEnterpriseEDR/ReleaseNotes/1_1_35.md:6]:
```suggestion
Fixed an issue in the **fetch-incidents** which used to return duplicate alerts.
```

**Comment by ShirleyDenkberg** [comment]
@RosenbergYehuda Doc review completed.

**Review by RosenbergYehuda — APPROVED** [review]
Awesome!!

**Pull request merged**